ec2_group: add rule description support - fixes #29040

s-hertel opened the pull request:
##### SUMMARY
Allow rule descriptions to be created and modified. Fixes #29040.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/ec2_group.py

##### ANSIBLE VERSION
```
2.5.0
```

s-hertel committed:
ec2_group: add support for rule descriptions.

s-hertel committed:
Document rule description feature and add an example using it.

s-hertel committed:
Fix removing rule descriptions.

s-hertel committed:
Add integration tests to verify adding/modifying/removing rule descriptions works as expected.

s-hertel commented:
Tests are passing locally. I need to add permissions for UpdateSecurityGroupRuleDescriptionsIngress and UpdateSecurityGroupRuleDescriptionsEgress.

willthames commented:
@s-hertel if you need extra permissions adding, can you also update hacking/aws_config/testing_policies/ec2-policy.json too

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -481,12 +497,32 @@ def rules_expand_sources(rules):
+def update_rules_description(module, client, rule_type, group_id, ip_permissions):
+    try:
+        if rule_type == "in":
+            client.update_security_group_rule_descriptions_ingress(GroupId=group_id, IpPermissions=[ip_permissions])
```
This could presumably fail with AttributeError if a user hasn't recently updated botocore.

s-hertel replied:
Thanks, I have addressed this in 7061e623ca34799693f213a7625bc19e1ac5681b. There might be a more elegant solution though, and I'm not sure if I should be making it a global - perhaps I should be passing it around explicitly. Let me know if you have a preference please.

I have also added to ec2-policy.json.

s-hertel committed:
Add permissions to hacking/aws_config/testing_policies/ec2-policy.json for updating ingress and egress rule descriptions.

s-hertel committed:
ec2_group: add backwards compatibility with older versions of botocore for rule descriptions.

s-hertel committed:
Add compatibility with older version of botocore for ec2_group integration tests.

s-hertel committed:
ec2_group: move HAS_RULE_DESCRIPTION to be checked first.

s-hertel committed:
Make requested change

s-hertel commented:
@willthames Is this what you wanted? 71eb1054fa4c461027bd1bfa864b6801e63ef5e8
I'm kind of unhappy with it because now client is being passed everywhere.

willthames commented:
@s-hertel, I think the problem is that you're checking whether the client can do something in a lot of places where it shouldn't be (and I should have picked up on that earlier, sorry).

The simpler fix is probably to check just after client creation whether the client supports rule descriptions - if it doesn't, and any rules have descriptions, fail at that point. Then in the serialization functions, only set the rule description if it exists.

s-hertel committed:
Pass around a variable instead of client

s-hertel committed:
Make sure has_rule_description defaults to None

s-hertel committed:
Fail if rule_desc is in any ingress/egress rules and the the botocore version < 1.7.2

s-hertel commented:
@willthames Thanks for looking this. When you have a chance will you review again?

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -515,12 +556,15 @@ def authorize_ip(type, changed, client, group, groupRules,
                   'ToPort': rule['to_port'],
                   'UserIdGroupPairs': [{'GroupId': group_id}]}
 
+    if has_rule_description:
```
you don't need `has_rule_description` for this method (or `serialize_ip_grant`).

There are two cases:
* There are no rule descriptions
* There are rule descriptions and botocore supports that (otherwise `fail_json` was called before we got here)

On that basis, you can do this here instead:

```
if 'rule_desc' in rule:
    permission['UserIdGroupPairs'][0]['Description'] = rule['rule_desc']
```

willthames commented on lib/ansible/modules/cloud/amazon/ec2_group.py:
```diff
@@ -489,16 +505,41 @@ def rules_expand_sources(rules):
         if rule_id in groupRules:
+
+            # update the rule description
+            if has_rule_description:
```
This can be `if 'rule_desc' in rule` and then you don't need to pass `has_rule_description` (see full reasons in my other comment)

s-hertel committed:
Remove unnecessary variable

s-hertel commented:
@willthames Thanks for bearing with me :) I think this is getting there. I removed the unnecessary variable and fixed the comparisons (in case they provide `rule_desc: ` which will be None, but needs to be compared/used as an empty string.

s-hertel committed:
Fix indentation for changed=True when updating rule descriptions.

s-hertel committed:
minor refactor to remove duplicate code

s-hertel committed:
add missing parameter

willthames commented on hacking/aws_config/testing_policies/ec2-policy.json:
```diff
@@ -51,7 +51,9 @@
                 "ec2:RevokeSecurityGroupEgress",
```
We need ec2:DeleteTags too.

willthames commented:
With the addition of ec2:DeleteTags to the testing policies, and the line length fixes, this will be good to go.

s-hertel committed:
Fix pep8

s-hertel committed:
Update test policy.

willthames commented:
rebuild_merge
